### ISSUE TITLE
a query with a nested subquery is too slow 

### ISSUE THREAD

**Issue by gaos1** [greenplum-db/gpdb#6553]
### Greenplum version or build
lastest master, same on gp 5 and gp4.
 PostgreSQL 9.4beta1 (Greenplum Database 6.0.0-alpha.0+d73a3711a0 build dev-oss) on x86_64-unknown-linux-gnu, compiled by gcc (GCC) 4.8.5 2015
0623 (Red Hat 4.8.5-28), 64-bit compiled on Dec 19 2018 17:04:10 (with assert checking)

### OS version and uname -a
Linux sdw 3.10.0-862.11.6.el7.x86_64

### Expected behavior
Need to find out the slow reason and see if we can improve its execution time. 

If we rewrote the query as shown below,  it can run very fast, but as commented by Pengzhou/Paul/Jesse, the query plan is actually wrong:
```
postgres=# explain analyze select (select id from (select * from t2) foo where name=t1.name) from t1;
                                                            QUERY PLAN                                                            
----------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 2:1  (slice1; segments: 2)  (cost=0.00..21889.00 rows=10000 width=4) (actual time=3.593..17.937 rows=5000 loops=2)
   ->  Seq Scan on t1  (cost=0.00..21889.00 rows=5000 width=4) (actual time=0.101..16.629 rows=2507 loops=2)
         SubPlan 1  (slice1; segments: 2)
           ->  Index Scan using bidx4 on t2  (cost=0.16..2.18 rows=1 width=4) (actual time=0.000..0.005 rows=1 loops=5015)
                 Index Cond: (name = t1.name)
 Planning time: 0.846 ms
   (slice0)    Executor memory: 354K bytes.
   (slice1)    Executor memory: 166K bytes avg x 2 workers, 166K bytes max (seg0).
 Memory used:  128000kB
 Optimizer: legacy query optimizer
 Execution time: 37.306 ms
(11 rows)
```
### Actual behavior
The query is quite slow and could take more than several hours, if both T1 and T2 have more than 10m rows
### Step to reproduce the behavior
* prepare table and data
```
drop table if exists t1;
drop table if exists t2;

create table t1(id int,name text) with (appendonly=false);
insert into t1 select i,i from generate_series(1,10000) i;
create table t2(id int,name text) with (appendonly=false);	
insert into t2 select i,i from generate_series(1,10000) i ;

create index bidx1 on t1(id);
create index bidx2 on t1(name);
create index bidx3 on t2(id);
create index bidx4 on t2(name);
```
* Run the query:
```
set random_page_cost=1;
set enable_nestloop_join=on;

## The following query is too slow
postgres=# explain analyze select (select id from t2 where t2.name=t1.name) from t1;
                                                                       QUERY PLAN                                                             
          
----------------------------------------------------------------------------------------------------------------------------------------------
----------
 Gather Motion 2:1  (slice2; segments: 2)  (cost=0.00..1390114.00 rows=10000 width=4) (actual time=660.696..4990.718 rows=5000 loops=2)
   ->  Seq Scan on t1  (cost=0.00..1390114.00 rows=5000 width=4) (actual time=5.822..4754.099 rows=2507 loops=2)
         SubPlan 1  (slice2; segments: 2)
           ->  Result  (cost=0.00..139.00 rows=1 width=4) (actual time=0.002..1.892 rows=1 loops=5015)
                 Filter: (t2.name = t1.name)
                 ->  Materialize  (cost=0.00..139.00 rows=1 width=4) (actual time=0.002..0.639 rows=9998 loops=5015)
                       ->  Broadcast Motion 2:2  (slice1; segments: 2)  (cost=0.00..139.00 rows=1 width=4) (actual time=0.244..3.667 rows=5000
 loops=2)
                             ->  Seq Scan on t2  (cost=0.00..139.00 rows=1 width=4) (actual time=0.253..0.525 rows=2507 loops=2)
 Planning time: 5.975 ms
   (slice0)    Executor memory: 322K bytes.
   (slice1)    Executor memory: 78K bytes avg x 2 workers, 78K bytes max (seg0).
 * (slice2)    Executor memory: 434K bytes avg x 2 workers, 434K bytes max (seg0).  Work_mem: 224K bytes max, 192K bytes wanted.
 Memory used:  128000kB
 Memory wanted:  692kB
 Optimizer: legacy query optimizer
 Execution time: 9994.533 ms
(16 rows)
```

**Comment by paul-guo-** [comment]
On merge code with pg9.4.19, it is not reproducible. Not sure PG  or GP code fixes that.

```
postgres=# explain analyze select (select id from (select * from t2) foo where name=t1.name) from t1;
                                                            QUERY PLAN

-------------------------------------------------------------------------------------------------------------------
---------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..21890.00 rows=10000 width=4) (actual time=5.237..23.577 rows
=5000 loops=2)
   ->  Seq Scan on t1  (cost=0.00..21890.00 rows=3334 width=4) (actual time=0.169..20.173 rows=1668 loops=2)
         SubPlan 1  (slice1; segments: 3)
           ->  Index Scan using bidx4 on t2  (cost=0.16..2.18 rows=1 width=4) (actual time=0.000..0.009 rows=1 loop
s=3337)
                 Index Cond: (name = t1.name)
 Planning time: 5.190 ms
   (slice0)    Executor memory: 350K bytes.
   (slice1)    Executor memory: 150K bytes avg x 3 workers, 150K bytes max (seg0).
 Memory used:  128000kB
 Optimizer: legacy query optimizer
 Execution time: 49.341 ms
(11 rows)
```

**Comment by pengzhout** [comment]
@gaos1 I'm afraid the first query
```
select (select id from (select * from t2) foo where name=t1.name) from t1;
```
is producing the wrong plan and wrong results in GPDB, comparing to the results of PG10.

**Comment by paul-guo-** [comment]
Pengzhou, I agree with you. Please keep this issue. @gaos1 Thanks.

**Comment by paul-guo-** [comment]
gpdb master output if using the distribution key for the condition. This output is suboptimal.
```
postgres=# explain select (select name from t2 where t2.id=t1.id) from t1;
                                                QUERY PLAN
-----------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..1400115.00 rows=10000 width=4)
   ->  Seq Scan on t1  (cost=0.00..1400115.00 rows=3334 width=4)
         SubPlan 1  (slice2; segments: 3)
           ->  Result  (cost=0.00..140.00 rows=1 width=4)
                 Filter: (t2.id = t1.id)
                 ->  Materialize  (cost=0.00..140.00 rows=1 width=4)
                       ->  Broadcast Motion 3:3  (slice1; segments: 3)  (cost=0.00..140.00 rows=1 width=4)
                             ->  Seq Scan on t2  (cost=0.00..140.00 rows=1 width=4)
 Planning time: 1.520 ms
 Optimizer: legacy query optimizer
(10 rows)
```

By the way, since we added motion node, the row number of seq scan is wrong and thus the cost is wrong.

**Comment by d** [comment]
The "expected" plan by original poster is wrong: if we don't broadcast `t2`, we will miss tuples from other segments that also match on `name`

**Comment by amp123321** [comment]
WHY DONT YALL STOP FUCKING CLICKING ON ALL MY GOD DAMN GOOGLE ADS??!!

**Comment by amp123321** [comment]
THAT WOULD BE REAL NICE!

**Comment by gaos1** [comment]
Thanks @pengzhout @paul-guo- @d 
So actually two issues here.  the "expected" query result and plan are wrong. I created a separate issue to track it: https://github.com/greenplum-db/gpdb/issues/6557

in this issue, let's focus if we can improve the performance of original correct query:
select (select name from t2 where t2.id=t1.id) from t1;

The query will take more than several hours if both T1 and T2 have more than 10m rows.  

I also updated the issue title and description according to your findings.
